### ISSUE TITLE
Nested workflow does not parallel when `if ... else` statement is used

### ISSUE THREAD

**Issue by gaow** [vatlab/sos#1374]
Here is the MWE:

```
[to_be_nested]
output: f"{_input:n}.out"
bash: expand = True
  sleep 3
  echo done > {_output}


[default]
input: '1.txt', '2.txt', '3.txt', '4.txt', '5.txt', group_by = 1
output: f"{_input:n}.out"
if True:
	sos_run('to_be_nested')
else:
	sos_run('to_be_nested')
```

To see the problem,

```
touch {1..5}.txt
sos run test.sos 
```

You see the nested workflows do not happen in parallel. If I remove the `if ... else` it will indeed run in parallel.

**Comment by BoPeng** [comment]
I guess you know the technical difficulties in running subworkflows in parallel if they are in the middle. Basically the entire step will have to wait and blocking a worker.

You however have a different case in that the subworkflows are at the end, so they could be run in parallel, and SoS will have to be a bit more clever to know that.

**Comment by BoPeng** [comment]
The relevant code is at https://github.com/vatlab/sos/blob/b1b937ee839dff252d3407238397589d573b79af/src/sos/step_executor.py#L1460 

and I am not sure if I can fix it.

**Comment by BoPeng** [comment]
Some basic clean up and I think this ticket will be WONTFIX.

The problem is not the running of workflows, which, with some work, we can identify the pattern and allow the workflows to execute in parallel. The problem is with the return value of `sos_run`. SoS uses the return value of `sos_run` to keep track of results from subworkflows. If `sos_run` is the last statement, SoS will complete the step and wait for the results from subworkflows. However, in the case of `if/else`, the statements will return `None` so no workflow information can be gathered...



```diff
diff --git a/src/sos/step_executor.py b/src/sos/step_executor.py
index 8d32a3ac..7b35c16a 100755
--- a/src/sos/step_executor.py
+++ b/src/sos/step_executor.py
@@ -345,12 +345,20 @@ def get_value_of_param(name, param_list, extra_dict={}):

 def is_sos_run_the_only_last_stmt(stmt):
     tree = ast.parse(stmt)
-    return len(tree.body) >= 1 and \
-        isinstance(tree.body[-1], ast.Expr) and \
-        isinstance(tree.body[-1].value, ast.Call) and \
-        hasattr(tree.body[-1].value.func, 'id') and \
-        tree.body[-1].value.func.id == 'sos_run' and \
-        len([x for x in ast.walk(tree) if isinstance(x, ast.Call) and hasattr(x.func, 'id') and x.func.id == 'sos_run']) == 1
+
+    def is_sos_run_expr(py_expr):
+        return isinstance(py_expr, ast.Expr) and isinstance(
+            py_expr.value, ast.Call) and hasattr(
+                py_expr.value.func, 'id') and py_expr.value.func.id == 'sos_run'
+
+    def num_sos_run_call(py_expr):
+        return len([
+            x for x in ast.walk(py_expr) if isinstance(x, ast.Call) and
+            hasattr(x.func, 'id') and x.func.id == 'sos_run'
+        ])
+
+    return len(tree.body) >= 1 and is_sos_run_expr(tree.body[-1]) and sum(
+        num_sos_run_call(x) for x in tree.body[:-1]) == 0


 class Base_Step_Executor:
```

**Comment by gaow** [comment]
Thank you @BoPeng. I was about to propose bringing back the `no_wait` if it helps but now I see the problem is return status. 

What would you suggest an alternative to implement that workflow though? Basically we have in our analysis a step where depending on the nature of data, we use different workflows. One is coded in R and the other in Python so we cannot do an if ... else inside one action. I hope there are some other syntax that can help?

**Comment by BoPeng** [comment]
Cannot you do

```
sos_run('workflow_R' if R else 'workflow_Py')
```
or even
```
if R:
  workflow_name =
  args = 
else:
  workflow_name = 
  args = 

sos_run(workflow_name, args)
```

**Comment by gaow** [comment]
Oh okay I forgot `sos_run` take kwargs which essentially is a dictionary, so the 2nd implementation can work. Thanks!

**Comment by gaow** [comment]
With this workflow we ended up having an error `Too many open files`. Again the worklow looks like

```
[default]
input: '1.txt', '2.txt', '3.txt', '4.txt', '5.txt', group_by = 1
output: f"{_input:n}.out"
if R:
  workflow_name =
  args = 
else:
  workflow_name = 
  args = 

sos_run(workflow_name, args)
```

When these workflows are executed not from nested workflow they do work. 

I am currently running with `-v4` to see if it provides more information. So no need to do anythign at this point unless you have an immediate thought what's going on.

**Comment by gaow** [comment]
@BoPeng here is the complete log:

[hybrid.log.gz](https://github.com/vatlab/sos/files/4702403/hybrid.log.gz)

and the last few lines are the errors:

```
DEBUG: EXECUTOR - Stop controller from 10339
DEBUG: EXECUTOR - disconntecting master
DEBUG: CONTROLLER - 10339 : closes socket with handler 17 (344 left)
DEBUG: CONTROLLER - 10339 : closes socket with handler 20 (343 left)
DEBUG: CONTROLLER - 10339 : closes socket with handler 23 (342 left)
DEBUG: CONTROLLER - controller stopped 10339
DEBUG: CONTROLLER - 10339 : closes socket with handler 39 (341 left)
DEBUG: CONTROLLER - 10339 : closes socket with handler 32 (340 left)
DEBUG: CONTROLLER - Disconnecting sockets from 10339
  File "/scratch/midway2/gaow/miniconda3/lib/python3.7/site-packages/sos/__main__.py", line 642, in cmd_run
    executor.run(args.__targets__, mode=config['run_mode'])
Traceback (most recent call last):
  File "/scratch/midway2/gaow/miniconda3/lib/python3.7/site-packages/sos/__main__.py", line 642, in cmd_run
    executor.run(args.__targets__, mode=config['run_mode'])
  File "/scratch/midway2/gaow/miniconda3/lib/python3.7/site-packages/sos/workflow_executor.py", line 347, in run
    return self.run_as_master(targets=targets, mode=mode)
  File "/scratch/midway2/gaow/miniconda3/lib/python3.7/site-packages/sos/workflow_executor.py", line 1649, in run_as_master
    raise exec_error
sos.executor_utils.ExecuteError: [hybrid]: Too many open files
[hybrid]: Exits with 1 pending step (hybrid_4)
ERROR: [hybrid]: Too many open files
[hybrid]: Exits with 1 pending step (hybrid_4)
```
